### PR TITLE
Fix #143: Add data creation CLI

### LINKED ISSUE #143: Containerise data index creation
For #140, the api-bot actually doesn't create the data index since the data lives with the machine hosting the LLM. To circumvent, containerise the data creation which gets stored to an Azure blob storage which gets ran with a GH action.

### PR DESCRIPTION
**Pull request by rchan26**
For #143:
- Create `reginald_create_index` CLI to create an index
- Create Dockerfile for `reginald_create_index`
- Split up data index creation from `LlamaIndex` to make a `DataIndexCreator` class
- Also add a `setup_service_context` helper function which gets called in both setting up LlamaIndex model and when creating the index on its own
- Some more logging with the parser
  - Created `get_env_var` helper function which just logs when it's trying to read a environment variable and says if it's able to find it. If not able to, default arguments are used
  - I made the default arguments in argparses lambda functions so that we only try to get environment variables if they are needed (a bit of a hack because otherwise, it'd log that it's trying to find an environment variable even if we pass in a value for the CLI argument)
  - `get_args` loops through the arguments and calls the arguments and calls any lambda functions to get the default arguments
- `k`, `chunk_size`, `chunk_overlap_ratio`, `num_output` are now able to be passed as CLI arguments in `reginald_run`, `reginald_run_api_llm` and `reginald_create_index`

### CODE REVIEW

**Comment by rchan26** [comment]
converting to draft as a few things to fix regarding how types of the default arguments. trying to fix this now

File: reginald/slack_bot/setup_bot.py [hunk 1]
```diff
@@ -83,19 +85,19 @@ def setup_slack_client(slack_bot: ApiBot | Bot) -> SocketModeClient:
...
         sys.exit(1)
 
     # initialize SocketModeClient with an app-level token + AsyncWebClient
     client = SocketModeClient(
         # this app-level token will be used only for establishing a connection
-        app_token=os.environ.get("SLACK_APP_TOKEN"),
+        app_token=get_env_var("SLACK_APP_TOKEN"),
         # you will be using this AsyncWebClient for performing Web API calls in listeners
-        web_client=AsyncWebClient(token=os.environ.get("SLACK_BOT_TOKEN")),
+        web_client=AsyncWebClient(token=get_env_var("SLACK_BOT_TOKEN")),
```
Comment (rwood-97) [inline reginald/slack_bot/setup_bot.py:100]:
```suggestion
    slack_app_token = get_env_variable("SLACK_APP_TOKEN")
    if slack_app_token is None:
        logging.error("SLACK_APP_TOKEN is not set")
        sys.exit(1)
    
    slack_bot_token = get_env_variable("SLACK_BOT_TOKEN")
    if slack_bot_token is None:
        logging.error("SLACK_BOT_TOKEN is not set")
        sys.exit(1)

    # initialize SocketModeClient with an app-level token + AsyncWebClient
    client = SocketModeClient(
        # this app-level token will be used only for establishing a connection
        app_token=slack_app_token,
        # you will be using this AsyncWebClient for performing Web API calls in listeners
        web_client=AsyncWebClient(token=slack_bot_token),
```

Comment (rwood-97) [reply reginald/slack_bot/setup_bot.py:100]: I don't really understand how it's done that highlighting lol sorry

File: reginald/parser_utils.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        the argument is actually used.
+
+        See `get_args` for parsing the arguments and obtaining any of the
+        environment variables if using default values.
+
+        Parameters
+        ----------
+        create_index_only : bool, optional
+            Whether or not to only include ones related to data index creation,
+            by default False
```
Comment (rwood-97) [inline reginald/parser_utils.py:25]:
```suggestion
            Whether or not to only include arguments related to data index creation,
```

Comment (rchan26) [reply reginald/parser_utils.py:25]: I'm a bit confused by this suggestion - would it mean repeating the "Whether or not to..." line?

Comment (rwood-97) [reply reginald/parser_utils.py:25]: I'm going to stop using vs code to do suggestions, I think it should say "Whether or not to only include arguments related to..."

**Comment by rwood-97** [comment]
Looks good - I changed the instances of `os.getenv` to `get_env_var` which means it prints the "getting SLACK_APP_TOKEN " bit twice (since it does actually get it twice) but I think that is okay?

**Pull request merged**